### DISCUSSION TITLE
how to cast value when using df.func.where?

### DISCUSSION THREAD

**Discussion by Hasham04** [vaexio/vaex#2128] [Q&A]
Hello,  I have a file that I am r.eading where one of the columns has values in scientific notation. like 4.503021124e+10, My goal is to output this file I am reading as a CSV.

I have tried using.
```
df[my_col]=df[my].astype('str')
df[my_col]=df.func.where(df[my_col].str.contains('e')==True, str(int(df[my_col])),df[my_col])

```
I basically want :

4.503021124e+10 to look like 45030211240

However, this gives me an error saying `int()` does not take an `Expression` which makes sense. Any help is greatly appreciated. Thank you.

**Comment by JovanVeljanoski** [comment]
Can you provide a small example file that you use for input? or describe its contents. Hard to offer suggestions without reproducing your exact usecase.

**Reply by Hasham04** [reply]
Hi, @JovanVeljanoski  thanks for replying. Here is sample data
[dummy_data.txt](https://github.com/vaexio/vaex/files/9185753/dummy_data.txt)

**Reply by JovanVeljanoski** [reply]
Ok thanks. 

This has nothing to do with vaex btw.. it is just a pandas option. Vaex (and pandas?) will display such long numbers with a scientific formation. To save them without a scientific notation to file (csv) just use

```python
# for a vaex dataframe
df.export_csv('../dummy_data_vaex.txt', index=False, float_format='{:f}'.format, encoding='utf-8')

# for a pandas dataframe
df.to_csv('../dummy_data_vaex.txt', index=False, float_format='{:f}'.format, encoding='utf-8')
```

This is all form the pandas docs btw. 

No need to cast anything or use `where` etc.. just a export formatting option.

**Reply by Hasham04** [reply]
Hi, @JovanVeljanoski thanks for your help. But my issue is that the CSV file I need to generate is over 100GB. I am converting the vaex data frame to an arrow table and then using the pyarrow CSV writer to write the table to csv. It is about 30x faster than pandas csv writer. Any way I can make the data like 4.503021124e+10 look like 45030211240 inside the data frame or table itself. Your help is greatly appreciated @JovanVeljanoski  and thanks for always replying.

**Reply by JovanVeljanoski** [reply]
Hey

Maybe you misunderstand things.. how the number appears in the csv is up to the "writer" (i.e. the pandas `to_csv` or the equivalent pyarrow function). How it looks like in vaex has no impact on that. So you need to find the settings in pyarrow that allows you to format the number as you see fit.

Btw, we do plan on supporting the pyarrow csv writer from vaex, and here is a general discussion about writing csv files faster from vaex: https://github.com/vaexio/vaex/issues/1687

Btw, how often do you plan on exporting 100GB csv files? If it is a one time thing.. you can consider just waiting it out.. if you need to do it often then yeah.. it is worth considering faster alternatives.

**Reply by Hasham04** [reply]
@JovanVeljanoski 

```
x = ['-129.66','-120','15','30','40','4.503021124e+10','5.503021124e+10']
df = vaex.from_arrays(x=x)
df['new_col']=df.func.where(df.x.str.contains('e')==True,df.x,'0')
df['new_col']=df['new_col'].astype('float')
df['new_col']=df['new_col'].astype('int64')
df['new_col']=df['new_col'].astype('str')

df['final']=df.func.where(df['x'].str.contains('e')==False,df['x'],df['new_col'])
```

![image](https://user-images.githubusercontent.com/69496736/181701574-25172cbf-b0df-42cc-a0b5-cc9835c18b11.png)

shouldn't the values in column final be  45030211584 and 55030210560  instead of  4.50302e+10 and 5.50302e+10. Any help is greatly appeeciated,

**Reply by JovanVeljanoski** [reply]
Why did you just ignore my answer before? As i said.. the way vaex represents data on screen when you print or display things is not related to how they are saved when you write them to a csv file..

How vaex prints stuff to screen is related to some internal formatting..

**Reply by Hasham04** [reply]
Hi Jovan, the final column in my last comment is of type string. why is a string being formatted to be in scientific notation? We are no longer talking about float formatting when generating CSV. A string should be represented as is correct. 

```
x = ['-129.66','-120','15','30','40','4.503021124e+10','5.503021124e+10']
df = vaex.from_arrays(x=x)
df['new_col']=df.func.where(df.x.str.contains('e')==True,df.x+'','0')
df['new_col']=df['new_col'].astype('float')
df['new_col']=df['new_col'].astype('int64')
df['new_col']=df['new_col'].astype('str')

df['final']=df.func.where(df['x'].str.contains('e')==False,df.x+'-',df['new_col']+'')
```

if I add '-' to the final column then the number is represented as I want it.

![image](https://user-images.githubusercontent.com/69496736/181801556-01d8af2f-8d7e-441d-bda5-996a96b1d486.png)

when I remove the `'-'` it goes back to being scientific notation.

**Reply by JovanVeljanoski** [reply]
"Why a strong is formatted in a scientific notation" is a different question then your original question in this thread. If you want to open a separate discussion on this, feel free to do so. 

<details>
  <summary>In short..</summary>
  
Say in stead of numbers you had text.. and one sample was a text entry with 10_000 characters. Do you want to see all of them? Would not be very nice.. you would probably see something like `"Lorem ipsum dolor est ..."`. Something similar is done in pandas as well. 

Here we recognize the string is integer so instead of doing `1234....` we can actually show you the whole thing!. 

At least this is my assumption. This part of vaex is quite complicated (for me), so if you are reallly interested in it.. you'll have to look in the code yourself..
</details>

I will repeat again. The way the strings are formatted when you do `print(df)` or `display(df)` (in Jupyter notebook for example), has nothing to do with the actual representation internally, or when you do operations to them.

For instance, for saving data to CSV, vaex uses pandas under the hood. So you can easily do `df.to_pandas_df()` and see that the strings are now formatted the way pandas formats them. And since pandas is the writer, you need to make sure you format your data in the way you want to. It would be similar when you use arrow. Again this is the same explanation as in my first reply.. does not matter if you are talking about strings, ints etc.. 

<img width="780" alt="image" src="https://user-images.githubusercontent.com/18574951/181850517-6ce089dd-3ee4-4a40-a3b5-a47add795959.png">
